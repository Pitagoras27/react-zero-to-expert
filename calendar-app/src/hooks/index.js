export * from "./useCalendarStore";
export * from "./useModalForm";
export * from "./useUiStore";
