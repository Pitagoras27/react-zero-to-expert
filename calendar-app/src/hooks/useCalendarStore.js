import { useDispatch, useSelector } from "react-redux";
import calendarApi from "../api/calendarApi";
import { onAddNewEvent, onDeleteEvent, onUpdateEvent, setActiveEvent } from "../store/calendar/calendarSlice";

export const useCalendarStore = () => {
  const { events, activeEvent } = useSelector((state) => state.calendar);
  const { user } = useSelector(state => state.auth);
  const dispatch = useDispatch();

  const onSetActiveEvent = (event) => {
    dispatch(setActiveEvent(event));
  };

  const startSavingEvent = async (calendarEvent) => {
    if (calendarEvent.id) {
      dispatch(onUpdateEvent({ ...calendarEvent }));
    } else {
      const { data } = await calendarApi.post('/events', calendarEvent);
      const { id } = data?.event;
      dispatch(onAddNewEvent({ ...calendarEvent, id, user }));
    }
  };

  const startDeleteEvent = () => {
    dispatch(onDeleteEvent());
  }

  return {
    events,
    activeEvent,
    hasActiveEvent: !!activeEvent,
    onSetActiveEvent,
    startSavingEvent,
    startDeleteEvent,
  };
};
