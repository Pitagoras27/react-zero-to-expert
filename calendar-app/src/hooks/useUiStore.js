import { useDispatch, useSelector } from "react-redux";
import { onModalDateClose, onModalDateOpen } from "../store";

export const useUiStore = () => {
  const { isDateModalOpen } = useSelector((state) => state.ui);
  const dispatch = useDispatch();

  const onModalOpen = () => {
    dispatch(onModalDateOpen());
  };

  const onCloseModal = () => {
    dispatch(onModalDateClose());
  };

  return {
    isDateModalOpen,
    onModalOpen,
    onCloseModal,
  };
};
