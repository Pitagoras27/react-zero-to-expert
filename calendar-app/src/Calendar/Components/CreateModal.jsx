import { addHours } from "date-fns";
import es from "date-fns/locale/es";
import DatePicker from "react-datepicker";
import Modal from "react-modal";

import "react-datepicker/dist/react-datepicker.css";
import { useModalForm, useUiStore } from "../../hooks";

const customStyles = {
  content: {
    top: "50%",
    left: "50%",
    right: "auto",
    bottom: "auto",
    marginRight: "-50%",
    transform: "translate(-50%, -50%)",
  },
};

Modal.setAppElement("#root");

export const CreateModal = () => {
  const { isDateModalOpen, onCloseModal } = useUiStore();
  const {
    title,
    notes,
    start,
    end,
    validationText,
    inputRef,
    onValueChange,
    onDateSelected,
    onSubmitEvent,
  } = useModalForm({
    title: "Carlor",
    notes: "Purification",
    start: new Date(),
    end: addHours(new Date(), 2),
  });

  return (
    <Modal
      isOpen={isDateModalOpen}
      onRequestClose={onCloseModal}
      style={customStyles}
      classModal="modal"
      overClassName="modal-fondo"
      closeTimeoutMS={200}
    >
      <h1> Nuevo evento </h1>
      <hr />
      <form className="container" onSubmit={onSubmitEvent}>
        <div className="form-group mb-2">
          <label>Fecha y hora inicio</label>
          <DatePicker
            selected={start}
            onChange={(date) => onDateSelected(date, "start")}
            className="form-control"
            dateFormat="MMMM d, yyyy h:mm aa"
            locale={es}
            timeCaption="hora"
            showTimeSelect
          />
        </div>

        <div className="form-group mb-2">
          <label>Fecha y hora fin</label>
          <DatePicker
            selected={end}
            onChange={(date) => onDateSelected(date, "end")}
            className="form-control"
            dateFormat="MMMM d, yyyy h:mm aa"
            minDate={start}
            locale={es}
            timeCaption="hora"
            showTimeSelect
          />
        </div>

        <hr />
        <div className="form-group mb-2">
          <label>Titulo y notas</label>
          <input
            type="text"
            className={`form-control ${validationText}`}
            placeholder="Título del evento"
            ref={inputRef}
            name="title"
            autoComplete="off"
            value={title}
            onChange={onValueChange}
          />
          <small id="emailHelp" className="form-text text-muted">
            Una descripción corta
          </small>
        </div>

        <div className="form-group mb-2">
          <textarea
            type="text"
            className="form-control"
            placeholder="Notas"
            rows="5"
            name="notes"
            value={notes}
            onChange={onValueChange}
          ></textarea>
          <small id="emailHelp" className="form-text text-muted">
            Información adicional
          </small>
        </div>

        <button type="submit" className="btn btn-outline-primary btn-block">
          <i className="far fa-save"></i>
          <span> Guardar</span>
        </button>
      </form>
    </Modal>
  );
};
