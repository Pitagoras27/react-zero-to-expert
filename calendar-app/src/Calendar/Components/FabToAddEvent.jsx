import { addHours } from "date-fns";
import { useCalendarStore, useUiStore } from "../../hooks";

export const FabToAddEvent = () => {
  const { onModalOpen } = useUiStore();
  const { onSetActiveEvent } = useCalendarStore();

  const handleAddNewEvent = () => {
    onSetActiveEvent({
      title: "",
      notes: "",
      start: new Date(),
      end: addHours(new Date(), 2),
      bgColor: "#fafafa",
      user: {
        _id: "",
        name: "",
      },
    });
    onModalOpen();
  };

  return (
    <button className="fab btn btn-primary" onClick={handleAddNewEvent}>
      <i className="fas fa-plus"></i>
    </button>
  );
};
