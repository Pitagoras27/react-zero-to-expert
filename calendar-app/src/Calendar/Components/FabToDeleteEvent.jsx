import { useCalendarStore, useUiStore } from "../../hooks";

export const FabToDeleteEvent = () => {

  const { hasActiveEvent, startDeleteEvent } = useCalendarStore();
  const { isDateModalOpen } = useUiStore();

  const handleDeleteEvent = () => {
    startDeleteEvent();
  };

  return (
    <button
      aria-label="delete-event"
      className="fab btn btn-danger fab-danger"
      onClick={handleDeleteEvent}
      style={{ display: (hasActiveEvent && !isDateModalOpen) ? '' : 'none'}}
    >
      <i className="fas fa-trash-alt"></i>
    </button>
  );
};
