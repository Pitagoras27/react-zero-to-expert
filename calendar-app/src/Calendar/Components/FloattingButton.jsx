import { addHours } from "date-fns";
import { useCalendarStore, useUiStore } from "../../hooks";

export const FloattingButton = () => {
  const { onModalOpen } = useUiStore();
  const { onSetActiveEvent } = useCalendarStore();

  const onAddNewEvent = () => {
    onModalOpen();
    onSetActiveEvent({
      title: "",
      notes: "",
      start: new Date(),
      end: addHours(new Date(), 2),
      bgColor: "#fafafa",
      user: {
        _id: "",
        name: "",
      },
    });
  };

  return (
    <button className="fab btn btn-primary" onClick={onAddNewEvent}>
      <i className="fas fa-plus"></i>
    </button>
  );
};
