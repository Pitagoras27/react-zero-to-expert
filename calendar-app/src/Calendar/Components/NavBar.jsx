import { useAuthStore } from "../../hooks";

export const NavBar = () => {
  const { startLogout } = useAuthStore();
  return (
    <div className='navbar navbar-dark bg-black mb-4 px-4'>
      <span className="navbar-brand">
        <i className="fas fa-calendar-alt"></i>&nbsp;
        Carlos
      </span>

      <button
        className="btn btn-outline-danger"
        onClick={startLogout}
      >
        <i className="fas fa-sign-out-alt"></i>
        <span style={{ marginLeft: '10px'}}>Logout</span>
      </button>
    </div>
  )
}