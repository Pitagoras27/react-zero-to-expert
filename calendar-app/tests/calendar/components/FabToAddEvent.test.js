import { fireEvent, render, screen } from "@testing-library/react";
import { FabToAddEvent } from "../../../src/Calendar/Components/FabToAddEvent";
import { useCalendarStore } from "../../../src/hooks/useCalendarStore";
import { useUiStore } from "../../../src/hooks/useUiStore";

jest.mock('../../../src/hooks/useUiStore');
jest.mock('../../../src/hooks/useCalendarStore')

describe('test in FabToAddEvent component', () => {
  const mockOnSetActiveEvent = jest.fn();
  const mockOnModalOpen = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test('should render  FabToAddEvent', () => {
    useCalendarStore.mockReturnValue({
      onSetActiveEvent: mockOnSetActiveEvent
    })
    useUiStore.mockReturnValue({
      onModalOpen: mockOnModalOpen
    })

    const { container } = render(<FabToAddEvent />);

    expect(container).toMatchSnapshot();
  });
  
  test('should onSetActiveEvent and onOpenModal have been called', () => {
    useCalendarStore.mockReturnValue({
      onSetActiveEvent: mockOnSetActiveEvent
    })
    useUiStore.mockReturnValue({
      onModalOpen: mockOnModalOpen
    })

    render(<FabToAddEvent />)

    const button = screen.getByRole('button');
    fireEvent.click(button);

    expect(mockOnSetActiveEvent).toHaveBeenCalled();
    expect(mockOnModalOpen).toHaveBeenCalled();
  })

})