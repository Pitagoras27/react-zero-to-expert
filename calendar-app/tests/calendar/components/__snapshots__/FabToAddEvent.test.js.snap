// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in FabToAddEvent component should render  FabToAddEvent 1`] = `
<div>
  <button
    class="fab btn btn-primary"
  >
    <i
      class="fas fa-plus"
    />
  </button>
</div>
`;
